### ISSUE TITLE
Python3.6 segfault calling `help()`

### ISSUE THREAD

**Issue by geigerj** [protocolbuffers/protobuf#2974]
Calling `help()` on protobuf messages results in a segfault with Python 3.6. Examples:

```
$ python
Python 3.6.1 (default, Apr 13 2017, 16:05:55)
[GCC 4.8.4] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from google.protobuf import duration_pb2
>>> help(duration_pb2.Duration)
Segmentation fault (core dumped)
```

```
$ python
Python 3.6.1 (default, Apr 13 2017, 16:05:55)
[GCC 4.8.4] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from google.protobuf import timestamp_pb2
>>> help(timestamp_pb2.Timestamp)
Segmentation fault (core dumped)
```

**Comment by geigerj** [comment]
cc: @lukesneeringer

**Comment by lukesneeringer** [comment]
This got worse in protobuf 3.3:

```python
(vision) lukesneeringer@lukesneeringer0:~/Code/Google/gax-python:master$ python
Python 3.6.0 (default, Jan  5 2017, 10:44:51) 
[GCC 4.8.4] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from google.protobuf import timestamp_pb2
*** Error in `python': free(): invalid size: 0x00007f2681eccae0 ***
Aborted (core dumped)
```

If I use `bpython`, I get the previous behavior (segfault when I try to issue `help`).

**Comment by lukesneeringer** [comment]
Has this gone anywhere? This is a really serious usability issue.

**Comment by lukesneeringer** [comment]
Ping.

**Comment by lukesneeringer** [comment]
Additional information: This appears not to happen on Mac OS X.

**Comment by lukesneeringer** [comment]
Okay, I can confirm that the version of Python is not the issue, but the operating system matters. `help()` works fine on Mac OS X, but segfaults on Ubuntu (Google corp machine).

Here's the segfault in 3.4:

```
lukesneeringer@lukesneeringer0:~$ mkvirtualenv --python=`which python3.4` protobuf-3.4
Running virtualenv with interpreter /usr/bin/python3.4
Using base prefix '/usr'
New python executable in /usr/local/google/home/lukesneeringer/.virtualenvs/protobuf-3.4/bin/python3.4
Also creating executable in /usr/local/google/home/lukesneeringer/.virtualenvs/protobuf-3.4/bin/python
Installing setuptools, pip, wheel...done.
virtualenvwrapper.user_scripts creating /usr/local/google/home/lukesneeringer/.virtualenvs/protobuf-3.4/bin/predeactivate
virtualenvwrapper.user_scripts creating /usr/local/google/home/lukesneeringer/.virtualenvs/protobuf-3.4/bin/postdeactivate
virtualenvwrapper.user_scripts creating /usr/local/google/home/lukesneeringer/.virtualenvs/protobuf-3.4/bin/preactivate
virtualenvwrapper.user_scripts creating /usr/local/google/home/lukesneeringer/.virtualenvs/protobuf-3.4/bin/postactivate
virtualenvwrapper.user_scripts creating /usr/local/google/home/lukesneeringer/.virtualenvs/protobuf-3.4/bin/get_env_details
(protobuf-3.4) lukesneeringer@lukesneeringer0:~$ pip install protobuf
Collecting protobuf
  Using cached protobuf-3.3.0-cp34-cp34m-manylinux1_x86_64.whl
Collecting six>=1.9 (from protobuf)
  Using cached six-1.10.0-py2.py3-none-any.whl
Requirement already satisfied: setuptools in ./.virtualenvs/protobuf-3.4/lib/python3.4/site-packages (from protobuf)
Installing collected packages: six, protobuf
Successfully installed protobuf-3.3.0 six-1.10.0
(protobuf-3.4) lukesneeringer@lukesneeringer0:~$ python
Python 3.4.3 (default, Nov 17 2016, 01:08:31) 
[GCC 4.8.4] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from google.protobuf import timestamp_pb2
>>> help(timestamp_pb2.Timestamp)
Segmentation fault (core dumped)
```

**Comment by lukesneeringer** [comment]
That said, you also should be running tests against 3.5 and 3.6. ;-)

**Comment by anandolee** [comment]
That is strange, I can't reproduce the error... Python 3.4 on my Google corp machine works well:

jieluo@jieluo:/tmp/anandolee/protobuf/python$ python3
Python 3.4.3 (default, Nov 17 2016, 01:08:31) 
[GCC 4.8.4] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from google.protobuf import duration_pb2
>>> help(duration_pb2.Duration)
#display the content and I entered 'q' for quit
>>>

**Comment by lukesneeringer** [comment]
I have all three 3.x Python versions installed from direct downloads from python.org (since we can not do the deadsnakes route). Maybe try that?

**Comment by jonparrott** [comment]
I tracked this down a bit more- this is happening only if protobuf is using its c extension. It happens because `help` uses `inspect` and `inspect` tries to access an `attribute` on `google.protobuf.pyext._message.Message`.

I can reproduce this easily with Python 3.6.2 with the following:

```
import google.protobuf.pyext._message
google.protobuf.pyext._message.Message._extensions_by_name
```

In fact, even just constructing the Message class causes a segfault:

```
google.protobuf.pyext._message.Message()
```

This may be a red herring though.

**Comment by dzen** [comment]
I just saw the segfault and attached a gdb producing this line:

```
Program received signal SIGSEGV, Segmentation fault.
google::protobuf::python::message_meta::GetExtensionsByName (self=0x7f80257543e0 <google::protobuf::python::CMessage_Type>, closure=<optimized out>) at google/protobuf/pyext/message.cc:349
349	google/protobuf/pyext/message.cc: No such file or directory.
```

**Comment by dzen** [comment]
Hi there,

can we change the title of this issue ?
It's not specific to `help()` it's more related when accessing to some attribute.
here is another issue:

```#5  0x00007f5bd2c60af9 in deallocate (this=<optimized out>, __p=<optimized out>) at /opt/rh/devtoolset-2/root/usr/include/c++/4.8.2/ext/new_allocator.h:110
#6  _M_put_node (this=<optimized out>, __p=<optimized out>) at /opt/rh/devtoolset-2/root/usr/include/c++/4.8.2/bits/stl_tree.h:374
#7  _M_destroy_node (this=<optimized out>, __p=<optimized out>) at /opt/rh/devtoolset-2/root/usr/include/c++/4.8.2/bits/stl_tree.h:396
#8  _M_erase_aux (__position=..., this=0x7f5bd2fc3160 <google::protobuf::python::interned_descriptors>) at /opt/rh/devtoolset-2/root/usr/include/c++/4.8.2/bits/stl_tree.h:1746
#9  erase (__position=..., this=0x7f5bd2fc3160 <google::protobuf::python::interned_descriptors>) at /opt/rh/devtoolset-2/root/usr/include/c++/4.8.2/bits/stl_tree.h:830
#10 _M_erase_aux (__last=..., __first=..., this=0x7f5bd2fc3160 <google::protobuf::python::interned_descriptors>) at /opt/rh/devtoolset-2/root/usr/include/c++/4.8.2/bits/stl_tree.h:1760
#11 erase (__last=..., __first=..., this=0x7f5bd2fc3160 <google::protobuf::python::interned_descriptors>) at /opt/rh/devtoolset-2/root/usr/include/c++/4.8.2/bits/stl_tree.h:848
#12 erase (__x=@0x7f5bc2688978: 0x3460170, this=0x7f5bd2fc3160 <google::protobuf::python::interned_descriptors>) at /opt/rh/devtoolset-2/root/usr/include/c++/4.8.2/bits/stl_tree.h:1771
#13 erase (__x=@0x7f5bc2688978: 0x3460170, this=0x7f5bd2fc3160 <google::protobuf::python::interned_descriptors>) at /opt/rh/devtoolset-2/root/usr/include/c++/4.8.2/bits/stl_map.h:727
#14 google::protobuf::python::descriptor::Dealloc (self=0x7f5bc2688968) at google/protobuf/pyext/descriptor.cc:370
#15 0x00007f5bdbe562ab in insertdict (value=<optimized out>, hash=<optimized out>, key=<optimized out>, mp=<optimized out>) at ../Objects/dictobject.c:823```

**Comment by orivej** [comment]
The crash in accessing `google.protobuf.pyext._message.Message._extensions_by_name` happens because metaclass getters [1] shadow class getters [2] and the intended dispatch from the latter to the former [3] does not run. This can be fixed by replacing `message_meta::Getters` with `0` at [1].

[1] https://github.com/google/protobuf/blob/099d99759101c295244c24d8954ec85b8ac65ce3/python/google/protobuf/pyext/message.cc#L443
[2] https://github.com/google/protobuf/blob/099d99759101c295244c24d8954ec85b8ac65ce3/python/google/protobuf/pyext/message.cc#L2849
[3] https://github.com/google/protobuf/blob/099d99759101c295244c24d8954ec85b8ac65ce3/python/google/protobuf/pyext/message.cc#L2615-L2616

**Comment by zmedico** [comment]
Same as https://github.com/google/protobuf/issues/2911, fixed by https://github.com/google/protobuf/pull/4365.

**Comment by anandolee** [comment]
It is fixed in google internal code by @amauryfa. Expect to see the fix in next release.

I will close this for clean up, feel free to reopen if it is still an issue after next release.

**Comment by manuel-koch** [comment]
I see the same problem on python 3.4.7 using latest protobuf 3.6.0 !?

I.e. trying the snippet from above
```
Python 3.4.7 (default, Dec 12 2017, 06:47:53)
[GCC 4.9.2] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import google.protobuf
>>> google.protobuf.__version__
'3.6.0'
>>> import google.protobuf.pyext._message
>>> google.protobuf.pyext._message.Message._extensions_by_name
Segmentation fault
```

**Comment by ccdavid** [comment]
Also seeing it on Ubuntu 18.04 with python 3.6.5 and protobuf 3.6.0 .

**Comment by craigcitro** [comment]
@anandolee did the fix make it into 3.6.1? (I don't see it having been ported.) Or should we be waiting on 3.7?

**Comment by anandolee** [comment]
No. We only sync with internal changes for major release. Minor releases are manually handled for some issues.

**Comment by craigcitro** [comment]
@anandolee two last questions:

1. Is there a rough timeline for 3.7?
1. Is it reasonable to leave this issue open, so folks finding it don't have to read all the way through to discover that it's not fixed here, only internally?

**Comment by anandolee** [comment]
We haven't start 3.7 yet but have Down-integrate from google internal to master branch a few days ago: 
https://github.com/google/protobuf/commit/6bbe197e9c1b6fc38cbdc45e3bf83fa7ced792a3#diff-23eeeb4347bdd26bfc6b7ee9a3b755dd

It should have the fix for this issue

**Comment by ptrebert** [comment]
For those like me looking for a way to get around this bug, and who cannot wait until the fix has made it into final releases (and who need this to work in a conda environment), the following is worth trying:

1. create/activate your conda env (no need to have `libprotobuf` or `protobuf` installed), say `protoenv`
2. download suitable release candidate, e.g., `wget https://github.com/protocolbuffers/protobuf/releases/download/v3.7.0rc2/protobuf-python-3.7.0-rc-2.tar.gz`
3. extract `tar xzvf protobuf-python-3.7.0-rc-2.tar.gz` and `cd` into `protobuf-3.7.0-rc-2` folder
4. run `./configure --prefix=/path/to/conda/envs/protoenv`
5. subsequently, run `make` and `make install`
6. after successful setup, the command `which protoc` should show the correct path: `/path/to/conda/envs/protoenv/bin/protoc` ; and this `protoc --version` should show `libprotoc 3.7.0`
7. now `cd` into `python` subfolder (i.e., `protobuf-3.7.0-rc-2/python`), and run `python setup.py build`, `python setup.py test` and finally `python setup.py install`
8. start python and try `import google.protobuf as pb` followed by `pb.__version__`, which should show `3.7.0rc2` (if the module is importable)

With that setup, I managed to compile the vg protobuf file and import the resulting `vg_pb2.py` in Python3.6 w/o getting a segfault.

**Comment by anandolee** [comment]
Closing it as 3.7.0 is already released
